src/System.Diagnostics.Tracing.Logger/src/System/Diagnostics/Tracing/Logger.cs

```diff
@@ -0,0 +1,206 @@
+        /// </summary>
+        /// <param name="messageName">The name for this logging message.  (e.g. ExceptionOccured, or DataItemRetrieved)</param>
+        /// <param name="level">The verbosity of this message.</param>
+        /// <param name="arguments">Any additional data.  Typically this is an anoymous type object.</param>
+        public virtual void Log(string messageName, LogLevel level, object arguments = null)
+        {
+            if (IsEnabled(level))
+            {
+                 Write(messageName, new LoggerArguments { Level = level, LoggerName = Name, Arguments = arguments });
+            }
+        }
+
+        /// <summary>
+        /// If the logger is enabled for 'level' then geneate a 'activityName'.Start message with the given argument.
+        /// It returns an IDisposable that when Disposed() will send a 'activityName'.Stop message with the SAME arguments 
+        /// object.  Thus the arguments object itself can be used as an ID that links the two together.
+        /// </summary>
+        public virtual IDisposable ActivityStart(string activityName, LogLevel level = LogLevel.Critical, object arguments = null)
+        {
+            var activity = new LoggerActivity(this, level, activityName, arguments);
+            if (IsEnabled(level))
+                Log(activityName + ".Start", level, arguments);
+            else
+                activity.Disposed = true;
+            return activity;
+        }
+
+        /// <summary>
+        /// Subscribe to the Logger, sending events to 'observer' for any logging that is enabled for verbosity 'level'.  
+        /// 
+        /// Note that in multi-subscriber cases you MAY get messages that have a more verbose level than you asked for.
+        /// You can filter them out yourself.   If there is only one subscriber (or all subscribers subscribe at the 
+        /// same verbosity.  It works fine.   
+        /// </summary>
+        public virtual IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer, Func<string, LogLevel, bool> filter)
+        {
+            _filter += filter;
+            return base.Subscribe(observer);
+        }
+
+        #region private
+        private sealed class LoggerActivity : IDisposable
+        {
+            public LoggerActivity(Logger logger, LogLevel level, string activityName, object arguments)
+            {
+                _logger = logger;
+                _level = level;
+                _activityName = activityName;
+                _arguments = arguments;
+            }
+            public bool Disposed;
+
+            public void Dispose()
+            {
+                if (!Disposed)
+                {
+                    Disposed = true;
+                    _logger.Log(_activityName + ".Stop", _level, _arguments);
+                }
+            }
+            Logger _logger;
+            LogLevel _level;
+            string _activityName;
+            object _arguments;
```

Comment:
Nit: to conform with style guidelines, these should all be explicitly annotated as ```private```